### PR TITLE
use jersey-media-json-jackson instead of jackson-jaxrs-json-provider

### PR DESCRIPTION
**Pull request by dcendents**
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

Required update to use `jersey-media-json-jackson` from `jersey2-api` instead of `jackson-jaxrs-json-provider` removed from `jackson2-api`

A release of the 3 plugins with the following PRs will need to be coordinated:

- https://github.com/jenkinsci/jersey2-api-plugin/pull/9
- https://github.com/jenkinsci/jackson2-api-plugin/pull/122
- https://github.com/jenkinsci/gitlab-plugin/pull/1255

### CODE REVIEW

**Comment by dcendents** [comment]
Here are snapshots of the 3 plugins for testing purposes:

- https://repo.jenkins-ci.org/snapshots/io/jenkins/plugins/jersey2-api/2.35-5-SNAPSHOT/jersey2-api-2.35-5-20220302.161331-2.hpi
- https://repo.jenkins-ci.org/snapshots/org/jenkins-ci/plugins/jackson2-api/2.13.1-999999-SNAPSHOT/jackson2-api-2.13.1-999999-20220302.151333-1.hpi
- https://repo.jenkins-ci.org/snapshots/org/jenkins-ci/plugins/gitlab-plugin/1.5.29-SNAPSHOT/gitlab-plugin-1.5.29-20220302.151459-1.hpi

File: pom.xml [hunk 1]
```diff
@@ -186,10 +186,12 @@
     <dependency>
       <groupId>io.jenkins.plugins</groupId>
       <artifactId>jersey2-api</artifactId>
+      <version>2.35-5-rc28.31f10d84d0ff</version> <!-- TODO https://github.com/jenkinsci/jersey2-api-plugin/pull/9 -->
```
Comment (basil) [inline pom.xml]: jenkinsci/jersey2-api-plugin#9

File: pom.xml [hunk 2]
```diff
@@ -186,10 +186,12 @@
...
     </dependency>
     <dependency>
       <groupId>org.jenkins-ci.plugins</groupId>
       <artifactId>jackson2-api</artifactId>
+      <version>2.13.1-257.vb_254f61c3a_2b_</version> <!-- TODO https://github.com/jenkinsci/jackson2-api-plugin/pull/122 -->
```
Comment (basil) [inline pom.xml]: jenkinsci/jackson2-api-plugin#122

**Comment by basil** [comment]
CI builds of all three PRs are passing:

- https://ci.jenkins.io/job/Plugins/job/jackson2-api-plugin/job/PR-122/
- https://ci.jenkins.io/job/Plugins/job/jersey2-api-plugin/job/PR-9/
- https://ci.jenkins.io/job/Plugins/job/gitlab-plugin/job/PR-1255/

Next step, I think, is to do some real-world testing of the latest incrementals.

**Comment by basil** [comment]
So I guess `org.glassfish.jersey.jackson.JacksonFeature` works with RESTEasy (which this repository uses rather than Jersey)? I believe that you've tested this, but it would be interesting to see a stack trace that contains RESTEasy, Jersey, and Jackson all together to convince myself that this cocktail works. :-)

**Comment by dcendents** [comment]
I'm not sure how to generate a stacktrace where you will see all of them listed together to convince you.

But here is some _redacted_ logging statements from jenkins after I triggered a webhook from gitlab:

```
Mar 02, 2022 3:46:18 PM INFO com.dabsquared.gitlabjenkins.webhook.GitLabWebHook getDynamic
WebHook called with url: /jenkins/project/***************
Mar 02, 2022 3:46:18 PM INFO com.dabsquared.gitlabjenkins.trigger.handler.AbstractWebHookTriggerHandler handle
*************** triggered for push.
Mar 02, 2022 3:46:29 PM INFO com.dabsquared.gitlabjenkins.util.CommitStatusUpdater retrieveGitlabProjectIds
Retrieving gitlab project ids
Mar 02, 2022 3:46:29 PM INFO com.dabsquared.gitlabjenkins.util.CommitStatusUpdater updateCommitStatus
Updating build '1503' to 'running'
Mar 02, 2022 3:46:41 PM INFO com.dabsquared.gitlabjenkins.util.CommitStatusUpdater retrieveGitlabProjectIds
Retrieving gitlab project ids
Mar 02, 2022 3:46:41 PM INFO com.dabsquared.gitlabjenkins.util.CommitStatusUpdater updateCommitStatus
Updating build '1503' to 'success'
```

image

image

**Comment by basil** [comment]
Sure, like I said, I believe that you've tested this. Glad that it's working. :-)

**Comment by dcendents** [comment]
Yes but our use of `gitlab-plugin` is limited to a few simple projects, we use the `gitlab-branch-source-plugin` more

**Comment by basil** [comment]
I guess this just makes me nervous because in principle we shouldn't need Jersey for anything other than the JAX-RS APIs (implemented by RESTEasy in this plugin). It seems unusual that this is working given that we're not even using Jersey in this repository. I'm not certain at first blush that we're ever invoking `JacksonFeature#configure`. Maybe this is just dead code?

If you feel up to it, try attaching a remote JVM debugger to a staging Jenkins controller and see if you can find flow of code (in the old version) where Jackson is called. It should be some code path starting in the JAX-RS API (implemented by RESTEasy) going through `jackson-jaxrs-json-provider` ending up in `jackson-core`. Once you find that code in `jackson-core`, try following the same flow of code in the new version. That code path should have shifted to RESTEasy -> `jersey-media-json-jackson` -> `jackson-core`.

**Comment by dcendents** [comment]
@basil first let me say I installed it on our production server and it worked flawlessly (except classpath issue with the `jira` plugin)

I tested with the debugger to step through the configuration and I can say it works exactly the same way with  `RESTEasy` and `jackson-jaxrs-json-provider` or `jersey-media-json-jackson`.

In the end what is really important is the class `com.dabsquared.gitlabjenkins.gitlab.JacksonConfig` which returns the `com.fasterxml.jackson.databind.ObjectMapper` to be used to (de)serialize json.

First, with the latest `gitlab-plugin` `1.5.28`:

- In `ResteasyGitLabClientBuilder` we register `com.fasterxml.jackson.jaxrs.json.JacksonJsonProvider`
- When `RESTEasy` wants to deserialize a json object, it asks its providers
  - `com.fasterxml.jackson.jaxrs.json.JacksonJsonProvider.hasMatchingMediaType` returns `true`
  - `com.fasterxml.jackson.jaxrs.base.ProviderBase.readFrom` calls `com.dabsquared.gitlabjenkins.gitlab.JacksonConfig.getContext(Class<?>)` which returns a `com.fasterxml.jackson.databind.ObjectMapper`

```
Daemon Thread [Running CpsFlowExecution[Owner[jenkins-gitlab-plugin-test/main/6:jenkins-gitlab-plugin-test/main #6]]] (Suspended (breakpoint at line 21 in JacksonConfig))	
	owns: ApacheHttpClient4Engine$2  (id=505)	
	JacksonConfig.getContext(Class<?>) line: 21	
	JacksonConfig.getContext(Class) line: 16	
	JacksonJsonProvider._locateMapperViaProvider(Class<?>, MediaType) line: 208	
	JacksonJsonProvider(ProviderBase<THIS,MAPPER,EP_CONFIG,MAPPER_CONFIG>).locateMapper(Class<?>, MediaType) line: 917	
	JacksonJsonProvider(ProviderBase<THIS,MAPPER,EP_CONFIG,MAPPER_CONFIG>)._endpointForReading(Class<Object>, Type, Annotation[], MediaType, MultivaluedMap<String,String>) line: 846	
	JacksonJsonProvider(ProviderBase<THIS,MAPPER,EP_CONFIG,MAPPER_CONFIG>).readFrom(Class<Object>, Type, Annotation[], MediaType, MultivaluedMap<String,String>, InputStream) line: 757	
	ClientReaderInterceptorContext(AbstractReaderInterceptorContext).readFrom(MessageBodyReader) line: 66	
	ClientReaderInterceptorContext(AbstractReaderInterceptorContext).proceed() line: 56	
	ApacheHttpClient4Engine$2(ClientResponse).readFrom(Class<T>, Type, MediaType, Annotation[]) line: 213	
	ApacheHttpClient4Engine$2(BuiltResponse).readEntity(Class<T>, Type, Annotation[]) line: 89	
	ApacheHttpClient4Engine$2(AbstractBuiltResponse).readEntity(GenericType<T>, Annotation[]) line: 256	
```

Now this time using the incremental builds:
1. `gitlab-plugin-1.5.29-rc1374.5c457d31d036.hpi`
1. `jackson2-api-2.13.1-257.vb_254f61c3a_2b_.hpi`
1. `jersey2-api-2.35-5-rc30.c420823129f6.hpi`


- In `ResteasyGitLabClientBuilder` we register `org.glassfish.jersey.jackson.JacksonFeature`
  - `JacksonFeature` registers `org.glassfish.jersey.jackson.internal.DefaultJacksonJaxbJsonProvider`
- When `RESTEasy` wants to deserialize a json object, it asks its providers
  - `org.glassfish.jersey.jackson.internal.jackson.jaxrs.json.JacksonJsonProvider.hasMatchingMediaType` returns `true`
  - `org.glassfish.jersey.jackson.internal.jackson.jaxrs.base.ProviderBase.readFrom` calls `com.dabsquared.gitlabjenkins.gitlab.JacksonConfig.getContext(Class<?>)` which returns a `com.fasterxml.jackson.databind.ObjectMapper`

```
Daemon Thread [Running CpsFlowExecution[Owner[jenkins-gitlab-plugin-test/main/7:jenkins-gitlab-plugin-test/main #7]]] (Suspended (breakpoint at line 21 in JacksonConfig))	
	owns: ApacheHttpClient4Engine$2  (id=529)	
	JacksonConfig.getContext(Class<?>) line: 21	
	JacksonConfig.getContext(Class) line: 16	
	DefaultJacksonJaxbJsonProvider(JacksonJsonProvider)._locateMapperViaProvider(Class<?>, MediaType) line: 216	
	DefaultJacksonJaxbJsonProvider(ProviderBase<THIS,MAPPER,EP_CONFIG,MAPPER_CONFIG>).locateMapper(Class<?>, MediaType) line: 951	
	DefaultJacksonJaxbJsonProvider(ProviderBase<THIS,MAPPER,EP_CONFIG,MAPPER_CONFIG>)._endpointForReading(Class<Object>, Type, Annotation[], MediaType, MultivaluedMap<String,String>) line: 880	
	DefaultJacksonJaxbJsonProvider(ProviderBase<THIS,MAPPER,EP_CONFIG,MAPPER_CONFIG>).readFrom(Class<Object>, Type, Annotation[], MediaType, MultivaluedMap<String,String>, InputStream) line: 788	
	ClientReaderInterceptorContext(AbstractReaderInterceptorContext).readFrom(MessageBodyReader) line: 66	
	ClientReaderInterceptorContext(AbstractReaderInterceptorContext).proceed() line: 56	
	ApacheHttpClient4Engine$2(ClientResponse).readFrom(Class<T>, Type, MediaType, Annotation[]) line: 213	
	ApacheHttpClient4Engine$2(BuiltResponse).readEntity(Class<T>, Type, Annotation[]) line: 89	
	ApacheHttpClient4Engine$2(AbstractBuiltResponse).readEntity(GenericType<T>, Annotation[]) line: 256	
```

I hope this makes you less nervous! ;-)

**Comment by basil** [comment]
Thank you Daniel, this is exactly the kind of reasoning I was trying to establish to prove to myself that the new code is working the same way as the old code was.

**Review by basil — APPROVED** [review]

File: pom.xml [hunk 3]
```diff
@@ -185,10 +185,12 @@
     <dependency>
       <groupId>io.jenkins.plugins</groupId>
       <artifactId>jersey2-api</artifactId>
+      <version>2.35-6-rc41.7dc9f104dff2</version> <!-- TODO https://github.com/jenkinsci/jersey2-api-plugin/pull/9 -->
```
Comment (basil) [inline pom.xml]:
```suggestion
      <version>2.35-6</version>/pull/9 -->
```

File: pom.xml [hunk 4]
```diff
@@ -185,10 +185,12 @@
...
     </dependency>
     <dependency>
       <groupId>org.jenkins-ci.plugins</groupId>
       <artifactId>jackson2-api</artifactId>
+      <version>2.13.2-264.v22b_4f3616f0c</version> <!-- TODO https://github.com/jenkinsci/jackson2-api-plugin/pull/122 -->
```
Comment (basil) [inline pom.xml]:
```suggestion
      <version>2.13.2.20220328-273.v11d70a_b_a_1a_52</version>
```

**Review by basil — APPROVED** [review]

File: pom.xml [hunk 5]
```diff
@@ -185,10 +185,12 @@
     <dependency>
       <groupId>io.jenkins.plugins</groupId>
       <artifactId>jersey2-api</artifactId>
+      <version>2.35-6</version>/pull/9 -->
```
Comment (basil) [inline pom.xml]:
```suggestion
      <version>2.35-6</version>
```

**Comment by basil** [comment]
Released in [1.5.30](https://github.com/jenkinsci/gitlab-plugin/releases/tag/gitlab-plugin-1.5.30).

**Pull request merged**